File: easy/2303_calculate_amount_paid_in_taxes/solution.cpp
#include <vector>

class Solution {
public:
  double calculateTax(std::vector<std::vector<int>> &brackets, int income) {
    double tax = 0;
    int lower = 0;

    for (const std::vector<int> &bracket : brackets) {
      int upper = bracket[0];
      int percent = bracket[1];

      int bracketIncome = std::min(upper - lower, income);
      lower = upper;

      tax += bracketIncome * percent / 100;
      income -= bracketIncome;

      if (income == 0)
        break;
    }

    return tax;
  }
};

File: easy/2303_calculate_amount_paid_in_taxes/Solution.java
class Solution {
  public double calculateTax(int[][] brackets, int income) {
    double tax = 0;

    int lower = 0;
    int remainingIncome = income;

    for (final int[] bracket : brackets) {
      final int upper = bracket[0];
      final int percent = bracket[1];

      final int bracketIncome = Math.min(upper - lower, remainingIncome);
      lower = upper;

      tax += bracketIncome * percent / 100.0;
      remainingIncome -= bracketIncome;

      if (remainingIncome == 0) {
        break;
      }
    }

    return tax;
  }
}
